### ISSUE TITLE
[RFC] Smart typescript types for application of aria-* props

### ISSUE THREAD

**Issue by Philzen** [chakra-ui/chakra-ui#6265]
### Description

Not a simple task, more of an epic, but potentially worthwhile: 

Implementing accessibility is not trivial, though it follows rather clear defined rules (see https://www.w3.org/TR/using-aria/). Developing against storybook helps a lot, but there may be some rules that could be directly baked into chakra components with typescript.

### Problem Statement/Justification

There are many inter-dependencies / more or less complex rules that are not obvious when trying to add a little 11y to your app, unless of course you are an expert and know the spec like the back of your hand.

For instance, regarding `aria-label` the spec says:
> Don't use `aria-label` or `aria-labelledby` on a span or div unless its given a role. 

A developer may be thinking: "Well i'm going to add an aria-label to this div that holds important visual information, which i'm going to represent it in the label text, that will surely increase a11y of my application". Now they added an aria-label in good faith, but it won't ever be useful unless a `role` is added to that div.

### Proposed Solution or API

This surely is possible using typescript, i believe via union interfaces for example (see this [excellent article](https://artsy.github.io/blog/2018/11/21/conditional-types-in-typescript/#a-real-use-case-for-distributive-conditional-types)).

The current `aria-*` props resolve to `AriaAttributes` in the `@types/react` package. These would need to be
1. broken down to a multitude of interfaces that encapsulate the rules as union types
2. those would then need to be applied to the chakra component code, depending on the type of html element they render.

As i am realizing the aria-props are not implemented in Chakra itself, i assume the breaking down of rules and grouping into separate interface is an out-of-scope project – however at the same time i believe somebody must have done this already somewhere. If there is an full-blown engine in storybook that can validate via javascript runtime code, somewhere out there a project must be working on implementing these rules for compile-time usage. That would solve task 1.

Not sure if there isn't also way to automate the annotation process described by task 2.

### Alternatives

As mentioned, storybook provides excellent a11y feedback – but only after the code has been typed and rendered.
Also, there must be a multitude of linters / CI-tools that can verify generated markup for a11y rules, but again, that comes too late.

I am yet to discover a solution that can do this at compile / parsing time via typescript or some other form of tokenization mechanism while writing the code in the IDE.
